song-server/src/main/java/bio/overture/song/server/security/JWTTokenConverter.java

```diff
@@ -0,0 +1,89 @@
+package bio.overture.song.server.security;
+
+import static bio.overture.song.core.exceptions.ServerErrors.FORBIDDEN_TOKEN;
+import static bio.overture.song.core.exceptions.ServerException.buildServerException;
+import static bio.overture.song.core.utils.Joiners.WHITESPACE;
+import static bio.overture.song.server.oauth.ExpiringOauth2Authentication.from;
+import static java.lang.System.currentTimeMillis;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+import bio.overture.song.server.oauth.ExpiringOauth2Authentication;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import lombok.NonNull;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+import org.springframework.security.oauth2.provider.OAuth2Authentication;
+import org.springframework.security.oauth2.provider.token.store.JwtAccessTokenConverter;
+
+@Slf4j
+public class JWTTokenConverter extends JwtAccessTokenConverter {
+
+  private static final String CONTEXT = "context";
+  private static final String SCOPE = "scope";
+  private static final String EXP = "exp";
+  private static final String CONTEXT_SCOPE_FIELD_NAME = CONTEXT + "." + SCOPE;
+
+  @SneakyThrows
+  public JWTTokenConverter(@NonNull String publicKey) {
+    super();
+    this.setVerifierKey(publicKey);
+    this.afterPropertiesSet();
+  }
+
+  @Override
+  public OAuth2Authentication extractAuthentication(Map<String, ?> map) {
+    // TODO: rtisma --- this is a hack since EGO does not implement jwts correctly
+    val mutatedMap = mutateMap(map);
+
+    val expirationTimestamp = parseExpirationTimestamp(map);
+    val secondsUntilExpiry = calcSecondsUntilExpiry(expirationTimestamp);
+    ExpiringOauth2Authentication authentication =
+        from(super.extractAuthentication(mutatedMap), secondsUntilExpiry);
+
+    // TODO: rtisma --- this is also a hack. the resourceIds maps to the "aud" field. This should be
+    // empty inorder for the OAuth2AuthenticationManager to process properly
+    authentication.getOAuth2Request().getResourceIds().clear();
+
+    return authentication;
+  }
+
+  private static long calcSecondsUntilExpiry(Long expirationTimestamp) {
+    val diff = expirationTimestamp - MILLISECONDS.toSeconds(currentTimeMillis());
+    return diff > 0 ? diff : 0;
```

Comment:
Generally this is fine, and you are fine in this situation, but you can get into trouble when boxing and widening happen in the wrong order. I would personally use `0L` whenever I can because when using `val` you are boxing primitives. Again, you are fine in these situations, just something to think about.